src/scipp/core/shape.py

```diff
@@ -8,22 +8,43 @@
 
 from .._scipp import core as _cpp
 from ._cpp_wrapper_util import call_func as _call_cpp_func
+from ._sizes import _parse_dims_shape_sizes
 from ..typing import VariableLike
 
 
-def broadcast(x: _cpp.Variable, dims: Union[List[str], Tuple[str, ...]],
-              shape: Sequence[int]) -> _cpp.Variable:
-    """Broadcast a variable.
+def broadcast(
+    x: VariableLike,
+    dims: Optional[Union[List[str], Tuple[str, ...]]] = None,
+    shape: Optional[Sequence[int]] = None,
+    sizes: Optional[Dict[str, int]] = None,
+) -> _cpp.Variable:
+    """Broadcast a Variable or a DataArray.
+    If the input is a DataArray, coordinates and attributes are shallow-copied
+    and masks are deep copied.
 
     Note that scipp operations broadcast automatically, so using this function
     directly is rarely required.
 
-    :param x: Variable to broadcast.
-    :param dims: List of new dimensions.
-    :param shape: New extents in each dimension.
-    :return: New variable with requested dimension labels and shape.
+    :param x: Variable or DataArray to broadcast.
+    :param dims: Optional (if sizes is specified), list of new dimensions.
+    :param shape: Optional (if sizes is specified), new extents in each dimension.
+    :param sizes: Optional, new dimension labels to sizes map.
+    :return: New Variable or DataArray with requested dimension labels and shape.
     """
-    return _call_cpp_func(_cpp.broadcast, x, dims, shape)
+    sizes = _parse_dims_shape_sizes(dims=dims, shape=shape, sizes=sizes)
+    if isinstance(x, _cpp.Variable):
+        return _call_cpp_func(_cpp.broadcast, x, sizes["dims"], sizes["shape"])
+    elif isinstance(x, _cpp.DataArray):
+        return _cpp.DataArray(data=_call_cpp_func(_cpp.broadcast, x.data, sizes["dims"],
+                                                  sizes["shape"]),
+                              coords={c: coord
+                                      for c, coord in x.coords.items()},
```

Comment:
Unrelated: We should finally extend the constructors so you can just pass `coords=other.coords`.